data/sensors_recent.php

```diff
@@ -166,33 +182,68 @@ function gateway_button($gw_id, $gw_data = null) {
 EOF;
 	}
 
+
 	function output_cell($rowspan, $data) {
 		echo("  <td rowspan=\"$rowspan\"> " . $data . "</td>\n");
 	}
 
+
+	// main loop:
 	while($row = $result->fetch_array(MYSQLI_ASSOC)) {
 		if ($row['message']) {
 			$message = json_decode($row['message'], true);
-			$metadata = $message['metadata'];
 
-			if (!array_key_exists('gateways', $metadata)) {
-				$gateways = [];
+			if ($row['source'] == 'ttn.v2') {
+				$metadata = $message['metadata'];
 
-				// Convert TTN staging to production format
-				foreach ($metadata as $meta) {
-					$meta['gtw_id'] = 'eui-' . strtolower($meta['gateway_eui']);
-					$meta['snr'] = $meta['lsnr'];
-					$gateways[] = $meta;
+				// fix old v2 record-layout:
+				if (!array_key_exists('gateways', $metadata)) {
+					$gateways = [];
+	
+					// Convert TTN staging to production format
+					foreach ($metadata as $meta) {
+						$meta['gtw_id'] = 'eui-' . strtolower($meta['gateway_eui']);
+						$meta['snr'] = $meta['lsnr'];
+						$gateways[] = $meta;
+					}
+	
+					$metadata = [
+						'gateways' => $gateways,
+						'frequency' => $metadata[0]['frequency'],
+						'data_rate' => $metadata[0]['datarate'],
+						'coding_rate' => $metadata[0]['codingrate'],
+					];
 				}
-
-				$metadata = [
-					'gateways' => $gateways,
-					'frequency' => $metadata[0]['frequency'],
-					'data_rate' => $metadata[0]['datarate'],
-					'coding_rate' => $metadata[0]['codingrate'],
-				];
-			}
-
+			} else if ($row['source'] == 'ttn.v3') {
+				$metadata = $message['uplink_message']['rx_metadata'];
+				$settings = $message['uplink_message']['settings'];
+	
+				if (!array_key_exists('gateway_ids', $metadata)) {
```

Comment:
Volgens mij is deze if bij ttn.v3 niet meer relevant, toch? Ik denk dat die key nooit bestaat (dus if altijd true), dus die kan gewoon weg?